### PR TITLE
Return a FxIndexSet in is_late_bound query.

### PR DESCRIPTION
**Pull request by cjgillot**
Fixes https://github.com/rust-lang/rust/issues/98890
r? @cuviper

### CODE REVIEW

File: compiler/rustc_middle/src/middle/resolve_lifetime.rs [hunk 1]
```diff
@@ -64,7 +64,7 @@ pub struct ResolveLifetimes {
     /// Set of lifetime def ids that are late-bound; a region can
     /// be late-bound if (a) it does NOT appear in a where-clause and
     /// (b) it DOES appear in the arguments.
-    pub late_bound: FxHashMap<LocalDefId, FxHashSet<LocalDefId>>,
+    pub late_bound: FxHashMap<LocalDefId, FxIndexSet<LocalDefId>>,
```
Comment (cuviper) [inline compiler/rustc_middle/src/middle/resolve_lifetime.rs:67]: Is that value constructed in a deterministic order though? AFAICS no, `fn convert_named_region_map` uses the iteration order of `named_region_map.late_bound`, which is a `HirIdSet = FxHashSet`.

(The test case does repeatedly pass for me though...)

**Comment by apiraino** [comment]
Visiting during T-compiler meeting.

Stable backport decision delayed waiting for this [comment](https://github.com/rust-lang/rust/pull/99219#discussion_r920512225) to be addressed. T-team discussion on [Zulip](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202022-07-14/near/289597827)

File: compiler/rustc_resolve/src/late/lifetimes.rs [hunk 2]
```diff
@@ -482,6 +482,11 @@ fn convert_named_region_map(tcx: TyCtxt<'_>, named_region_map: NamedRegionMap) -
         let def_id = tcx.hir().local_def_id(hir_id);
         map.insert(def_id);
     }
+    for (_, late_bound) in &mut rl.late_bound {
+        late_bound.sort_by(|&a, &b| {
```
Comment (wesleywiser) [inline compiler/rustc_resolve/src/late/lifetimes.rs:486]: nit: I think it would be good to leave a comment here explaining why we're doing this sort

**Review by wesleywiser — APPROVED** [review]

**Review by cuviper — APPROVED** [review]
`r+`, but we'll add it to a stable release PR rather than merging it here.

**Comment by Mark-Simulacrum** [comment]
Merging since this was bundled into #99299.

**Pull request closed without merging**